Add EDF Teleinfo sensor

nlamirault opened the pull request:
This a new PR for the teleinfo sensor (Old PR: https://github.com/home-assistant/home-assistant/pull/10603)

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

## Description:

Add a new sensor for EDF Teleinfo, the data stream available on french
electric meters. (Description in french here)
It provides plenty of informations:

meter identification
subscribed power
subscribed contract
immediate power consumption
etc ...

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4382

## Example entry for `configuration.yaml` (if applicable):
```yaml
- platform: teleinfo
  name: "edf"
  device: "/dev/ttyUSB0"

- platform: template
  sensors:
    teleinfo_hchc:
      value_template: '{{ states.sensor.edf.attributes["HCHC"] }}'
      unit_of_measurement: 'Wh'

- platform: template
  sensors:
    teleinfo_hchp:
      value_template: '{{ states.sensor.edf.attributes["HCHP"] }}'
      unit_of_measurement: 'Wh'

- platform: template
  sensors:
    teleinfo_hhphc:
      value_template: '{{ states.sensor.edf.attributes["HHPHC"] }}'
      unit_of_measurement: ''

- platform: template
  sensors:
    teleinfo_iinst:
      value_template: '{{ states.sensor.edf.attributes["IINST"] }}'
      unit_of_measurement: 'A'

- platform: template
  sensors:
    teleinfo_imax:
      value_template: '{{ states.sensor.edf.attributes["IMAX"] }}'
      unit_of_measurement: 'A'

- platform: template
  sensors:
    teleinfo_papp:
      value_template: '{{ states.sensor.edf.attributes["PAPP"] }}'
      unit_of_measurement: 'vA'

- platform: template
  sensors:
    teleinfo_ptec:
      value_template: '{{ states.sensor.edf.attributes["PTEC"] }}'
      unit_of_measurement: ''

- platform: template
  sensors:
    teleinfo_isousc:
      value_template: '{{ states.sensor.edf.attributes["ISOUSC"] }}'
      unit_of_measurement: 'A'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54


```bash
$ tox -epy36 
Results (185.88s):
    3574 passed
     100 skipped
____________________________________________________________________________________ summary ____________________________________________________________________________________
  py36: commands succeeded
  congratulations :)
```

nlamirault committed:
Add: EDF Teleinfo sensor

Add a new sensor for EDF Teleinfo, the data stream available on french
electric meters. (Description in french here)
It provides plenty of informations:

meter identification
subscribed power
subscribed contract
immediate power consumption
etc ...

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault commented:
According to my TravisCI tests, unit tests seems OK but not pylint. Any idea ?

https://travis-ci.org/nlamirault/home-assistant/builds/373135602

dgomes commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,160 @@
+
+    @property
+    def teleinfo(self):
+        """Retour Teleinfo object."""
```
Retour -> Return ? (believe Retour is french for Return)

dgomes commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,160 @@
+        if self._teleinfo is not None:
+            self._teleinfo.close()
+            self._teleinfo = None
+        return
```
return is doing nothing here...

dgomes commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,160 @@
+        self._teleinfo.close()
+        return self._frame
+
+    def _stop(self, event):
```
this method is never called...

nlamirault committed:
Fix: typo due to PR code review

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'teleinfo'
```
No domain needed, it's a platform.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+
+DOMAIN = 'teleinfo'
+
+CONF_DEVICE = 'device'
```
This is already present in `const.py`. Import that please.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+DEFAULT_DEVICE = '/dev/ttyUSB0'
+DEFAULT_NAME = 'teleinfo'
+
+DEFAULT_NAME = 'Teleinfo'
```
Duplicate.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=1)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
```
If it's required then there should be no default. Otherwise it's optional.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+
+    except exceptions.KylinSerialError as err:
+        _LOGGER.error("Can't connect to teleinfo device: %s", str(err))
+        return False
```
Just return.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+
+    if not teleinfo_data.update():
+        _LOGGER.critical("Can't retrieve Teleinfo from device")
+        return False
```
Dito.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+        """Initialize the sensor."""
+        self._name = name
+        self._unit_of_measurement = None
+        self._state = STATE_UNKNOWN
```
Set it to `None` and let Home Assistant handle it.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+        return self._attributes
+
+    @property
+    def unit_of_measurement(self):
```
Not needed if there are no units of measurement.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+        return self._unit_of_measurement
+
+    @property
+    def attribution(self):
```
Sensor don't have this property. Add it to the attributes.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+        if not self._data.frame:
+            _LOGGER.warning("Don't receive energy data from Teleinfo!")
+            return
+        _LOGGER.info("Frame read: %s", self._data.frame)
```
Set this to debug.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from Teleinfo device."""
+        _LOGGER.info("Read teleinfo informations")
```
This is too verbose. It is already logged.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+        """Get the latest data from Teleinfo device."""
+        _LOGGER.info("Read teleinfo informations")
+        import kylin
+        self._teleinfo = kylin.Kylin(port=self._device, timeout=2)
```
Move this to the platform setup and pass the object to `TeleinfoData`.

fabaff commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,159 @@
+            if 'ADCO' == info['name']:
+                self._state = self._attributes['ADCO']
+        _LOGGER.debug(
+            "Sensor: state=%s attributes=%s", self._state, self._attributes)
```
Move the error checking and the logging to `TeleinfoData`.

nlamirault committed:
Fix: refactoring with code review

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Fix: pylint

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Update: lint

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

fabaff commented:
There are still comments which weren't addressed and a lint error.

nlamirault committed:
Update: move Attribution to sensor attributes

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Fix: lint

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Update: use Kylin 0.5.0

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault commented:
@fabaff i don't really know how fix this pylint error : 

```
 ImportError: No module named 'serial'
```

According to the documentation : https://pythonhosted.org/pyserial/shortintro.html#opening-serial-ports

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+
+    teleinfo_data = None
+    try:
+        import kylin
```
The import can probably be moved to the beginning of the function where we import the exceptions. Or is there a reason to put the import within the try... except?

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+        teleinfo_data = TeleinfoData(hass, teleinfo)
+
+    except exceptions.KylinSerialError as err:
+        return False
```
`setup_platform` shouldn't return anything, so just `return` here, or even better `raise PlatformNotReady`.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+class TeleinfoData(object):
+    """Get the latest data from Teleinfo."""
+
+    def __init__(self, hass, teleinfo):
```
`hass` isn't used.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+        self._teleinfo = teleinfo
+
+    @property
+    def teleinfo(self):
```
This property isn't needed. Just set `self.teleinfo` directly in init.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+        return self._teleinfo
+
+    @property
+    def frame(self):
```
Same as above.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+        self._frame = self._teleinfo.readframe()
+        self._teleinfo.close()
+        if not self._frame:
+            _LOGGER.warning("Don't receive energy data from Teleinfo!")
```
This log message uses weird grammar. Maybe:
`"No energy data received from Teleinfo"`

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+            _LOGGER.debug("Frame read: %s", self._data.frame)
+            for info in self._data.frame:
+                if info['name'] in TELEINFO_AVAILABLE_VALUES:
+                    self._attributes[info['name']] = int(info['value'])
```
What are the keys in the attributes here? Home assistant prefers to use lowercase snakecase keys in the state attributes.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+                else:
+                    self._attributes[info['name']] = info['value']
+                if 'ADCO' == info['name']:
+                    self._state = self._attributes['ADCO']
```
See above.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+                    self._attributes[info['name']] = info['value']
+                if 'ADCO' == info['name']:
+                    self._state = self._attributes['ADCO']
+            _LOGGER.debug("Sensor: %s %s", self._state, self._attributes)
```
Remove this. It will be logged by the core at info level.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+            return None
+        return self._frame
+
+    def _stop(self, event):
```
This is never called.

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,148 @@
+DEFAULT_DEVICE = '/dev/ttyUSB0'
+DEFAULT_NAME = 'teleinfo'
+
+DATA_TELEINFO = "data_teleinfo"
```
This isn't used.

MartinHjelmare commented on tests/components/sensor/test_teleinfo.py:
```diff
@@ -0,0 +1,123 @@
+    def setUp(self):
+        """Set up things to run when tests begin."""
+        self.hass = get_test_home_assistant()
+        self.config = VALID_CONFIG_NAME
```
This isn't used.

MartinHjelmare commented:
@nlamirault the pylint error is due to that you import the `kylin` package in your setup.py file in the `kylin` library, when you import the version.
https://github.com/nlamirault/kylin/blob/master/setup.py#L20

You should never import the package in the setup.py file if you have external requirements not in standard library, since other requirements are not installed yet when the setup.py file is read. If you want to single source the version in your library, please follow one of the recommended approaches here, other than number 6:
https://packaging.python.org/guides/single-sourcing-package-version/

Also, please don't pin `pyserial` to a specific version. It's ok to require a minimum version but don't pin it, and try to set as loose requirements as possible.
https://github.com/nlamirault/kylin/blob/master/setup.py#L46

nlamirault committed:
Update: new release of Kylin

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

MartinHjelmare commented:
Please run `script/gen_requirements.py` again and solve the remaining code review comments.

nlamirault committed:
Update: Set state to None. HA will handle it

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Update: Not needed if there are no units of measurement

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Update: unused code

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

balloob commented:
This can be merged once requirements issue has been solved and linting issue resolved. Looks like your library has some trouble installing.

nlamirault commented:
@balloob i don't understand the `pylint` error on `pyserial`. 

```
tox -r
GLOB sdist-make: /home/nlamirault/Perso/kylin/setup.py
py36 recreate: /home/nlamirault/Perso/kylin/.tox/py36
py36 installdeps: -r/home/nlamirault/Perso/kylin/requirements.txt, -r/home/nlamirault/Perso/kylin/requirements-test.txt
py36 inst: /home/nlamirault/Perso/kylin/.tox/dist/kylin-0.6.0.zip
py36 installed: astroid==2.0.4,coverage==4.5.1,flake8==3.4.1,isort==4.3.4,kylin==0.6.0,lazy-object-proxy==1.3.1,mccabe==0.6.1,py==1.6.0,pycodestyle==2.3.1,pyflakes==1.5.0,pylint==2.1.1,pyserial==3.4,pytest==3.2.3,pytest-cov==2.5.1,pytest-sugar==0.9.0,six==1.11.0,termcolor==1.1.0,typed-ast==1.1.0,wrapt==1.10.11
py36 runtests: PYTHONHASHSEED='3627596937'
py36 runtests: commands[0] | py.test --ignore=venv
Test session starts (platform: linux, Python 3.6.6, pytest 3.2.3, pytest-sugar 0.9.0)
rootdir: /home/nlamirault/Perso/kylin, inifile:
plugins: sugar-0.9.0, cov-2.5.1

 kylin/tests/test_kylin.py ✓                                                                                                                                      100% ██████████

Results (0.06s):
       1 passed
pep8 recreate: /home/nlamirault/Perso/kylin/.tox/pep8
pep8 installdeps: -r/home/nlamirault/Perso/kylin/requirements.txt, -r/home/nlamirault/Perso/kylin/requirements-test.txt
pep8 inst: /home/nlamirault/Perso/kylin/.tox/dist/kylin-0.6.0.zip
pep8 installed: astroid==2.0.4,coverage==4.5.1,flake8==3.4.1,isort==4.3.4,kylin==0.6.0,lazy-object-proxy==1.3.1,mccabe==0.6.1,py==1.6.0,pycodestyle==2.3.1,pyflakes==1.5.0,pylint==2.1.1,pyserial==3.4,pytest==3.2.3,pytest-cov==2.5.1,pytest-sugar==0.9.0,six==1.11.0,termcolor==1.1.0,typed-ast==1.1.0,wrapt==1.10.11
pep8 runtests: PYTHONHASHSEED='3627596937'
pep8 runtests: commands[0] | flake8
pylint recreate: /home/nlamirault/Perso/kylin/.tox/pylint
pylint installdeps: -r/home/nlamirault/Perso/kylin/requirements.txt, -r/home/nlamirault/Perso/kylin/requirements-test.txt
pylint inst: /home/nlamirault/Perso/kylin/.tox/dist/kylin-0.6.0.zip
pylint installed: astroid==2.0.4,coverage==4.5.1,flake8==3.4.1,isort==4.3.4,kylin==0.6.0,lazy-object-proxy==1.3.1,mccabe==0.6.1,py==1.6.0,pycodestyle==2.3.1,pyflakes==1.5.0,pylint==2.1.1,pyserial==3.4,pytest==3.2.3,pytest-cov==2.5.1,pytest-sugar==0.9.0,six==1.11.0,termcolor==1.1.0,typed-ast==1.1.0,wrapt==1.10.11
pylint runtests: PYTHONHASHSEED='3627596937'
pylint runtests: commands[0] | pylint kylin

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 9.81/10, +0.19)

____________________________________________________________________________________ summary ____________________________________________________________________________________
  py36: commands succeeded
  pep8: commands succeeded
  pylint: commands succeeded
  congratulations :)
```
@balloob do you have any idea ?

MartinHjelmare commented:
Run `script/gen_requirements.py` again and commit the changes.

MartinHjelmare commented:
Also add the module to `.coveragerc`.

nlamirault committed:
Update: use Kylin v0.6.0 and teleinfo sensor on coveragerc

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Fix: python lint code

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Fix: typo

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault commented:
@MartinHjelmare i think coverage decreased due to adding file to *.coveragerc* ?

MartinHjelmare commented:
I forgot there were tests. Then we can remove the module from `.coveragerc`. Make sure we cover all code with tests.

nlamirault committed:
Active Teleinfo sensor

nlamirault committed:
Add: Unit tests for teleinfo sensor

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Update: typo

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault committed:
Update: documentation

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

nlamirault commented:
@MartinHjelmare i add some unit tests

MartinHjelmare commented on homeassistant/components/sensor/teleinfo.py:
```diff
@@ -0,0 +1,140 @@
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from Teleinfo device."""
+        self._teleinfo.open()
```
Why do we need to open and close the serial connection on each update? I'm thinking we can open the connection in setup_platform and close it when home assistant stops.

nlamirault committed:
Update: HA stop handler and serial data

Use the HA Stop event to close the connection to the serial device.
Refactoring open/close serial connection
Update log level from critical to error.

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

balloob commented:
There are still a lot of comments that haven't been addressed and PR hasn't been updated in over a month. Going to close it.

guiguid commented:
Hi, is this PR include in HA ?
I'm very interested by this PR.
My Python skill can't help....

nlamirault commented:
@guiguid . No. I do not have the time currently to change it. I use teleinfo sensor on my HA configuration : 

- https://github.com/nlamirault/home-assistant-configuration/tree/master/custom_components/sensor
- https://github.com/nlamirault/home-assistant-configuration/blob/master/ui-lovelace.yaml#L191

esciara commented:
Hi. I am interested in this PR and would like to contribute. I am knowledgeable in Python but less so in contributing to other people's PRs. Hence my question: can I contribute to this particular PR or is it just not possible to avoid creating a new one ?

emontnemery commented:
You need to open a new PR. 
This is a really good guide of you're new to GitHub: https://gist.github.com/Chaser324/ce0505fbed06b947d962

For home assistant, also read up here: https://developers.home-assistant.io/docs/en/development_index.html

BaQs commented:
The component is broken since 0.92: 

> Integration teleinfo not found when trying to verify its sensor platform.

 I see in other similar components it needs a manifest and an init file. Not enough skilled in python to propose a change. If anybody wants to fix it... :)

esciara commented:
@emontnemery thanks. I will give it a shot. 

I see there are quite a few comments related to changes in the way HA works. Could you point me to well formed sensor code (or any other good code) on which I could get inspiration to do the job properly. A real example is worth a thousand words. 

Thx. Em
